ssz_serialization.nim

```diff
@@ -0,0 +1,250 @@
+    writeSeq(w, asSeq value)
+  elif value is BitList:
+    # ATTENTION! We can reuse `writeSeq` only as long as our BitList type is implemented
+    # to internally match the binary representation of SSZ BitLists in memory.
+    writeSeq(w, bytes value)
+  elif value is object|tuple|array:
+    trs "WRITING OBJECT OR ARRAY"
+    var ctx = beginRecord(w, type value)
+    enumerateSubFields(value, field):
+      writeField w, ctx, astToStr(field), field
+    endRecord w, ctx
+  else:
+    unsupported type(value)
+
+proc writeValue*(w: var SszWriter, x: auto) {.gcsafe, raises: [Defect, IOError].} =
+  mixin toSszType
+  type T = type toSszType(x)
+
+  when isFixedSize(T):
+    w.stream.writeFixedSized toSszType(x)
+  else:
+    w.writeVarSizeType toSszType(x)
+
+func sszSize*(value: auto): int {.gcsafe, raises: [Defect].}
+
+func sszSizeForVarSizeList[T](value: openArray[T]): int =
+  result = len(value) * offsetSize
+  for elem in value:
+    result += sszSize(toSszType elem)
+
+func sszSize*(value: auto): int {.gcsafe, raises: [Defect].} =
+  mixin toSszType
+  type T = type toSszType(value)
+
+  when isFixedSize(T):
+    anonConst fixedPortionSize(T)
+
+  elif T is array|List|HashList|HashArray:
+    type E = ElemType(T)
+    when isFixedSize(E):
+      len(value) * anonConst(fixedPortionSize(E))
+    elif T is HashArray:
+      sszSizeForVarSizeList(value.data)
+    elif T is array:
+      sszSizeForVarSizeList(value)
+    else:
+      sszSizeForVarSizeList(asSeq value)
+
+  elif T is BitList:
+    return len(bytes(value))
+
+  elif T is SingleMemberUnion:
+    sszSize(toSszType value.value) + 1
+
+  elif T is object|tuple:
+    result = anonConst fixedPortionSize(T)
+    enumInstanceSerializedFields(value, _{.used.}, field):
+      type FieldType = type toSszType(field)
+      when not isFixedSize(FieldType):
+        result += sszSize(toSszType field)
+
+  else:
+    unsupported T
+
+proc writeValue*[T](w: var SszWriter, x: SizePrefixed[T]) {.raises: [Defect, IOError].} =
+  var cursor = w.stream.delayVarSizeWrite(Leb128.maxLen(uint64))
+  let initPos = w.stream.pos
+  w.writeValue T(x)
+  let length = toBytes(uint64(w.stream.pos - initPos), Leb128)
+  cursor.finalWrite length.toOpenArray()
+
+proc readValue*(r: var SszReader, val: var auto) {.
+    raises: [Defect, MalformedSszError, SszSizeMismatchError, IOError].} =
+  mixin readSszBytes
+  type T = type val
+  when isFixedSize(T):
+    const minimalSize = fixedPortionSize(T)
+    if r.stream.readable(minimalSize):
+      readSszBytes(r.stream.read(minimalSize), val)
+    else:
+      raise newException(MalformedSszError, "SSZ input of insufficient size")
+  else:
+    # TODO(zah) Read the fixed portion first and precisely measure the
+    # size of the dynamic portion to consume the right number of bytes.
+    readSszBytes(r.stream.read(r.stream.len.get), val)
+
+proc readSszBytes*[T](data: openArray[byte], val: var T) {.
```

Comment:
this was really intended as an eth2 specific thing - but one thought I've had is to remove nim-serialization completely from the process so as to have a stand-alone SSZ reading library that operates on `openArray` - then `nim-ser` can be glued on top for any .. uh .. advanced use cases should they materialize.